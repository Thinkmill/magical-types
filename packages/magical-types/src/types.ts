import { typeParameter } from "@babel/types";

export type ObjectNode = {
  type: "Object";
  name: string | null;
  properties: Array<Property>;
};

export type ClassNode = {
  type: "Class";
  name: string | null;
  typeParameters: Array<MagicalNode>;
  thisNode: MagicalNode | null;
  properties: Array<Property>;
};

export type TypeParameterNode = { type: "TypeParameter"; value: string };

export type MagicalNode =
  | { type: "Intrinsic"; value: string }
  | { type: "StringLiteral"; value: string }
  | { type: "NumberLiteral"; value: number }
  | { type: "Array"; value: MagicalNode }
  | { type: "ReadonlyArray"; value: MagicalNode }
  | { type: "Union"; types: Array<MagicalNode>; name: string | null }
  | { type: "Intersection"; types: Array<MagicalNode> }
  | {
      type: "Function";
      return: MagicalNode;
      parameters: Array<Parameter>;
      typeParameters: Array<TypeParameterNode>;
    }
  | TypeParameterNode
  | { type: "Tuple"; value: Array<MagicalNode> }
  | ClassNode
  | ObjectNode;

export type Property = { key: string; value: MagicalNode };

type Parameter = {
  name: string;
  type: MagicalNode;
  required: boolean;
};
