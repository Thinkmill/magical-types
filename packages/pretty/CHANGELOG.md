# @magical-types/pretty

## 0.3.0

### Minor Changes

- [`571541e`](https://github.com/mitchellhamilton/magical-types/commit/571541e536b19dd825994e505584fc93addc8980) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Remove the heading prop from PropTypes

## 0.2.0

### Minor Changes

- [`bcf7960`](https://github.com/mitchellhamilton/magical-types/commit/bcf7960fbe1da21c5c394ee3c707894966df4dfd) [#26](https://github.com/mitchellhamilton/magical-types/pull/26) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support lazy loading nodes

### Patch Changes

- Updated dependencies [[`bcf7960`](https://github.com/mitchellhamilton/magical-types/commit/bcf7960fbe1da21c5c394ee3c707894966df4dfd)]:
  - @magical-types/utils@0.1.0

## 0.1.0

### Minor Changes

- [`d5f98c2`](https://github.com/mitchellhamilton/magical-types/commit/d5f98c2d28d81dcee07ba403c9336e7439f559c3) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Improve displaying some types

## 0.0.1

### Patch Changes

- [`86f1bf4`](https://github.com/mitchellhamilton/magical-types/commit/86f1bf4d4b17ec791bffd0a9841311152ebb5614) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Initial release
