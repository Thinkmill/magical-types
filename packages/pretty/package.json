{
  "name": "@magical-types/pretty",
  "version": "0.0.0",
  "main": "dist/pretty.cjs.js",
  "module": "dist/pretty.esm.js"
}
