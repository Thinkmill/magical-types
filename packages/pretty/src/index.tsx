import React, { useEffect, useReducer, useMemo, useState } from "react";
/** @jsx jsx */
import { css, jsx } from "@emotion/core";
import ReactMarkdown from "react-markdown";
import {
  MagicalNode,
  ObjectNode,
  ClassNode,
  TypeParameterNode,
  SignatureNode,
  PositionedMagicalNode,
  Parameter,
  LazyNode
} from "@magical-types/types";
import {
  Type,
  Indent,
  StringType,
  TypeMeta,
  Description,
  Required
} from "./pretty-proptypes/components";
import { colors, gridSize } from "./pretty-proptypes/components/constants";
import AddBrackets, {
  bracketStyle,
  PathExpansionContext
} from "./pretty-proptypes/PrettyConvert/AddBrackets";
import { getChildPositionedMagicalNodes, flatMap } from "@magical-types/utils";
import PropTypeHeading from "./pretty-proptypes/Prop/Heading";
import PropsWrapper from "./pretty-proptypes/Props/Wrapper";

const Arrow = () => (
  <span
    css={css`
      color: ${colors.G500};
    `}
  >
    {" => "}
  </span>
);

const TypeMinWidth = (props: React.HTMLAttributes<HTMLSpanElement>) => (
  <span
    css={css`
      display: inline-block;
      min-width: 60px;
    `}
    {...props}
  />
);

function Properties({
  node,
  path
}: {
  node: ClassNode;
  path: Array<number | string>;
}) {
  return (
    <Indent>
      {node.properties.map((prop, index) => {
        return (
          <div key={index}>
            {prop.description !== "" && (
              <div>
                <ReactMarkdown source={prop.description} />
              </div>
            )}
            <TypeMinWidth>
              <Type>{prop.key}</Type>
            </TypeMinWidth>
            {prop.required ? <Required> required</Required> : null}{" "}
            <RenderNode
              node={prop.value}
              path={path.concat("properties", index, "value")}
            />
          </div>
        );
      })}
    </Indent>
  );
}

function PrettyObject({
  node,
  path
}: {
  node: ObjectNode;
  path: Array<number | string>;
}) {
  return (
    <Indent>
      {node.constructSignatures.map((signature, index) => {
        return (
          <PrettySignature
            key={index}
            node={signature}
            funcType="colon"
            path={path.concat("constructSignatures", index)}
            type="construct"
          />
        );
      })}
      {node.callSignatures.map((signature, index) => {
        return (
          <PrettySignature
            key={index}
            node={signature}
            funcType="colon"
            path={path.concat("callSignatures", index)}
            type="call"
          />
        );
      })}
      {node.properties.map((prop, index) => {
        return (
          <div key={index}>
            {prop.description !== "" && (
              <div>
                <ReactMarkdown source={prop.description} />
              </div>
            )}
            <TypeMinWidth>
              <Type>{prop.key + (prop.required ? "" : "?")}</Type>
            </TypeMinWidth>
            <RenderNode
              node={prop.value}
              path={path.concat("properties", index, "value")}
            />
          </div>
        );
      })}
    </Indent>
  );
}

let cache = new Map<
  TypeParameterNode,
  { listeners: Array<(...args: any) => any>; value: boolean }
>();

// yes, i know this is bad

function PrettyTypeParameter({ node }: { node: TypeParameterNode }) {
  let [, forceUpdate] = useReducer(() => ({}), {});
  let val = cache.get(node);
  if (!val) {
    val = { listeners: [], value: false };
    cache.set(node, val);
  }

  let workingValue = val;

  useEffect(() => {
    workingValue.listeners.push(forceUpdate);
    return () => {
      workingValue.listeners = workingValue.listeners.filter(
        x => x !== forceUpdate
      );
    };
  }, []);

  return (
    <Type
      css={bracketStyle({ isHovered: val.value })}
      onMouseEnter={() => {
        workingValue.value = true;
        workingValue.listeners.forEach(listener => {
          listener();
        });
      }}
      onMouseLeave={() => {
        workingValue.value = false;
        workingValue.listeners.forEach(listener => {
          listener();
        });
      }}
    >
      {node.value}
    </Type>
  );
}

function Parameters({
  parameters,
  path
}: {
  parameters: Parameter[];
  path: Array<string | number>;
}) {
  return (
    <React.Fragment>
      {parameters.map((param, index, array) => (
        <React.Fragment key={index}>
          {param.name ? (
            <Type>
              {param.name}
              {param.required ? "" : "?"}:{" "}
            </Type>
          ) : (
            undefined
          )}
          <RenderNode
            node={param.type}
            path={path.concat("parameters", index, "type")}
          />
          {array.length - 1 === index ? "" : ", "}
        </React.Fragment>
      ))}
    </React.Fragment>
  );
}

function PrettySignature({
  node,
  path,
  type,
  funcType
}: {
  node: SignatureNode;
  path: Array<string | number>;
  type: "construct" | "call";
  funcType: "arrow" | "colon";
}) {
  return (
    <span>
      {type === "construct" ? "new " : ""}
      {node.typeParameters.length !== 0 && (
        <span>
          <span css={bracketStyle({ isHovered: false })}>{"<"}</span>
          {node.typeParameters.map((param, index, array) => (
            <React.Fragment key={index}>
              <RenderNode
                node={param}
                path={path.concat("typeParameters", index)}
              />
              {array.length - 1 === index ? "" : ", "}
            </React.Fragment>
          ))}
          <span css={bracketStyle({ isHovered: false })}>{">"}</span>
        </span>
      )}
      <AddBrackets nodes={node.parameters.map(x => x.type)}>
        <Parameters parameters={node.parameters} path={path} />
      </AddBrackets>
      {funcType === "arrow" ? (
        <Arrow />
      ) : (
        <span
          css={css`
            color: ${colors.G500};
          `}
        >
          {": "}
        </span>
      )}
      <AddBrackets nodes={[node.return]} openBracket="" closeBracket="">
        <RenderNode node={node.return} path={path.concat("return")} />
      </AddBrackets>
    </span>
  );
}

let RenderNode: (props: {
  node: MagicalNode;
  path: Array<string | number>;
}) => React.ReactElement = function RenderNode({ node, path }): any {
  if (path.length > 2000) {
    return "This node has a path of greater than 2000, this is mostly likely because of a bug in Magical Types";
  }
  switch (node.type) {
    case "Intrinsic": {
      return <Type>{node.value}</Type>;
    }
    case "StringLiteral": {
      return <StringType>"{node.value}"</StringType>;
    }
    case "NumberLiteral": {
      return <Type>{node.value}</Type>;
    }
    case "ReadonlyArray":
    case "Promise":
    case "Array": {
      let newPath = path.concat("value");

      return (
        <span>
          <TypeMeta>{node.type}</TypeMeta>
          <AddBrackets
            nodes={[node.value]}
            initialIsShown={newPath}
            openBracket="<"
            closeBracket=">"
            closedContent={
              (node as any).value.name ? (node as any).value.name : undefined
            }
          >
            <RenderNode path={newPath} node={node.value} />
          </AddBrackets>
        </span>
      );
    }
    case "Tuple": {
      return (
        <span>
          <TypeMeta>Tuple</TypeMeta>
          <AddBrackets
            nodes={node.value}
            initialIsShown={path}
            openBracket="["
            closeBracket="]"
          >
            {node.value.map((node, index, array) => (
              <React.Fragment key={index}>
                <RenderNode node={node} path={path.concat("value", index)} />
                {array.length - 1 === index ? "" : ", "}
              </React.Fragment>
            ))}
          </AddBrackets>
        </span>
      );
    }
    case "IndexedAccess": {
      return (
        <span>
          <div>
            <RenderNode node={node.object} path={path.concat("object")} />
          </div>
          <div>
            [<RenderNode node={node.index} path={path.concat("index")} />]
          </div>
        </span>
      );
    }
    case "TypeParameter": {
      return <PrettyTypeParameter node={node} />;
    }
    case "Union": {
      return (
        <span>
          <TypeMeta>
            {node.name === null ? "" : `${node.name} `}One of{" "}
          </TypeMeta>
          <AddBrackets
            nodes={node.types}
            initialIsShown={path}
            openBracket="<"
            closeBracket=">"
          >
            <Indent>
              {node.types.map((n, index, array) => (
                <div key={index}>
                  <RenderNode node={n} path={path.concat("types", index)} />
                  {array.length - 1 === index ? "" : ", "}
                </div>
              ))}
            </Indent>
          </AddBrackets>
        </span>
      );
    }
    case "Intersection": {
      let arr: Array<React.ReactNode> = [];
      node.types.forEach((type, index) => {
        arr.push(
          <span key={index}>
            <RenderNode node={type} path={path.concat("types", index)} />
          </span>
        );
        if (index < node.types.length - 1) {
          arr.push(<div key={`divider-${index}`}>&</div>);
        }
      });
      return arr;
    }
    case "Object": {
      if (
        node.callSignatures.length === 1 &&
        node.properties.length === 0 &&
        node.constructSignatures.length === 0
      ) {
        return (
          <PrettySignature
            node={node.callSignatures[0]}
            type="call"
            funcType="arrow"
            path={path.concat("callSignatures", 0)}
          />
        );
      }
      if (
        node.constructSignatures.length === 1 &&
        node.properties.length === 0 &&
        node.callSignatures.length === 0
      ) {
        return (
          <PrettySignature
            node={node.constructSignatures[0]}
            type="construct"
            funcType="arrow"
            path={path.concat("constructSignatures", 0)}
          />
        );
      }

      return (
        <span>
          <TypeMeta>{node.name}</TypeMeta>
          {node.callSignatures.length ||
          node.constructSignatures.length ||
          node.properties.length ? (
            <AddBrackets
              nodes={getChildPositionedMagicalNodes({
                node,
                path: [],
                depth: 0
              }).map(x => x.node)}
              initialIsShown={false}
              openBracket="{"
              closeBracket="}"
            >
              <PrettyObject path={path} node={node} />
            </AddBrackets>
          ) : (
            <span css={bracketStyle({ isHovered: false })}>{"{}"}</span>
          )}
        </span>
      );
    }
    case "Class": {
      return (
        <span>
          <TypeMeta>class {node.name}</TypeMeta>
          <AddBrackets
            nodes={null}
            initialIsShown={path}
            openBracket="{"
            closeBracket="}"
          >
            <Properties path={path} node={node} />
          </AddBrackets>
        </span>
      );
    }
    case "Conditional": {
      return (
        <span>
          <RenderNode node={node.check} path={path.concat("check")} />{" "}
          <TypeMeta>extends</TypeMeta>{" "}
          <RenderNode node={node.extends} path={path.concat("extends")} />{" "}
          <TypeMeta> ? </TypeMeta>
          <RenderNode node={node.true} path={path.concat("true")} />
          <TypeMeta> : </TypeMeta>
          <RenderNode node={node.false} path={path.concat("false")} />
        </span>
      );
    }
    case "Symbol": {
      return (
        <span>
          <TypeMeta>
            Symbol(
            <span css={{ color: colors.P300 }}>{node.name}</span>)
          </TypeMeta>
        </span>
      );
    }
    case "Lazy": {
      return <LazyNodeView node={node} path={path} />;
    }
    default: {
      let _thisMakesTypeScriptEnsureThatAllNodesAreSpecifiedHere: never = node;
    }
  }
};

function LazyNodeView({
  node,
  path
}: {
  node: LazyNode;
  path: Array<number | string>;
}) {
  let promise = node.loader();

  let value = node.value;
  let [, setState] = useState({});

  useEffect(() => {
    if (promise !== undefined) {
      promise.then(() => {
        setState({});
      });
    }
  }, [promise]);

  if (value !== undefined) {
    return <RenderNode node={value} path={path.concat("value")} />;
  }
  return <React.Fragment>Loading...</React.Fragment>;
}

// what are we doing here?
// first, what do we want to achieve?
// important note to understand this: magical nodes can contain circular references
// 1. we only want to expand each node by default a single time
// 2. the shallowest node and then first node at that depth is the one that should be expanded
// so, how do we implement this?
// we need to do a breadth first graph traversal
// Wikipedia explains what this is: https://en.wikipedia.org/wiki/Breadth-first_search

function getPathsThatShouldBeExpandedByDefault(rootNode: MagicalNode) {
  let pathsThatShouldBeExpandedByDefault = new Set<string>();

  // because of circular references, we don't want to visit a node more than once
  let visitedNodes = new Set<MagicalNode>();

  let queue: Array<PositionedMagicalNode> = [
    { node: rootNode, path: [], depth: 0 }
  ];

  while (queue.length) {
    let currentPositionedNode = queue.shift()!;

    visitedNodes.add(currentPositionedNode.node);
    if (
      ([
        "Object",
        "Union",
        "Array",
        "ReadonlyArray",
        "Tuple",
        "Class",
        "Promise",
        "Builtin"
      ] as Array<MagicalNode["type"]>).includes(currentPositionedNode.node.type)
    ) {
      pathsThatShouldBeExpandedByDefault.add(
        currentPositionedNode.path.join(":")
      );
    }
    // we don't want to open any nodes deeper than 5 nodes by default
    if (currentPositionedNode.depth < 3) {
      let childPositionedNodes = getChildPositionedMagicalNodes(
        currentPositionedNode
      ).filter(({ node }) => !visitedNodes.has(node));
      if (childPositionedNodes.length < 20 || currentPositionedNode.depth < 2) {
        queue.push(...childPositionedNodes);
      }
    }
  }
  return pathsThatShouldBeExpandedByDefault;
}

export let renderTypes = (node: MagicalNode) => {
  let pathsThatShouldBeExpandedByDefault = useMemo(() => {
    return getPathsThatShouldBeExpandedByDefault(node);
  }, [node]);

  return (
    <div
      css={{
        fontFamily:
          "source-code-pro,Menlo,Monaco,Consolas,Courier New,monospace"
      }}
    >
      <PathExpansionContext.Provider value={pathsThatShouldBeExpandedByDefault}>
        <RenderNode node={node} path={[]} />
      </PathExpansionContext.Provider>
    </div>
  );
};

const PropTypeWrapper = (props: { children: React.ReactNode }) => (
  <div
    css={css`
      margin-top: ${gridSize * 4}px;
    `}
    {...props}
  />
);

function simplifyIntersection(node: MagicalNode): MagicalNode {
  if (node.type !== "Intersection") {
    return node;
  }
  let types = node.types;
  if (types.every(value => value.type === "Object")) {
    let objectNodes = types as ObjectNode[];
    return {
      type: "Object",
      callSignatures: flatMap(objectNodes, x => x.callSignatures),
      constructSignatures: flatMap(objectNodes, x => x.callSignatures),
      properties: flatMap(objectNodes, x => x.properties),
      name: null,
      // TODO: fix this
      aliasTypeArguments: []
    };
  }
  return node;
}

export function Types({ node }: { node: MagicalNode }) {
  return renderTypes(node);
}

export function PropTypes({
  node,
  heading
}: {
  node: MagicalNode;
  heading?: string;
}) {
  node = simplifyIntersection(node);
  let pathsThatShouldBeExpandedByDefault = useMemo(() => {
    return getPathsThatShouldBeExpandedByDefault(node);
  }, [node]);
  if (node.type === "Object") {
    return (
      <PathExpansionContext.Provider value={pathsThatShouldBeExpandedByDefault}>
        <PropsWrapper heading={heading}>
          {node.properties.map((prop, index) => {
            return (
              <PropTypeWrapper key={index}>
                <PropTypeHeading name={prop.key} required={prop.required} />
                {prop.description && (
                  <Description>
                    <ReactMarkdown source={prop.description} />
                  </Description>
                )}
                <RenderNode
                  node={prop.value}
                  path={["properties", index, "value"]}
                />
              </PropTypeWrapper>
            );
          })}
        </PropsWrapper>
      </PathExpansionContext.Provider>
    );
  }
  return renderTypes(node);
}
