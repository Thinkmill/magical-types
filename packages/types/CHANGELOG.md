# @magical-types/types

## 0.1.2

### Patch Changes

- [`a127bc4`](https://github.com/Thinkmill/magical-types/commit/a127bc4a27334328a338ece03b82146ee87076c5) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add `name` to Intersection nodes

* [`a127bc4`](https://github.com/Thinkmill/magical-types/commit/a127bc4a27334328a338ece03b82146ee87076c5) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Change fields which were null on nodes to undefined so they're smaller when stringified

- [`915242f`](https://github.com/Thinkmill/magical-types/commit/915242fa35c2c6573c6fc50358f0ca9b6ec2cbc5) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add numberIndex and stringIndex to Object nodes

## 0.1.1

### Patch Changes

- [`3100747`](https://github.com/Thinkmill/magical-types/commit/31007479e4028efa67b383dc92208e9063335ab1) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add Error node type to handle unserialisable nodes without stopping everything from working

## 0.1.0

### Minor Changes

- [`bcf7960`](https://github.com/mitchellhamilton/magical-types/commit/bcf7960fbe1da21c5c394ee3c707894966df4dfd) [#26](https://github.com/mitchellhamilton/magical-types/pull/26) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Support lazy loading nodes

## 0.0.2

### Patch Changes

- [`6b9d0c2`](https://github.com/mitchellhamilton/magical-types/commit/6b9d0c2b479558aeb583b3ee7a9492748b9f6698) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Add `MagicalNodeWithIndexes` and `MagicalNodeIndex`

## 0.0.1

### Patch Changes

- [`86f1bf4`](https://github.com/mitchellhamilton/magical-types/commit/86f1bf4d4b17ec791bffd0a9841311152ebb5614) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Initial release
