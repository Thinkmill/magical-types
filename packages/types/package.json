{
  "name": "@magical-types/types",
  "version": "0.0.1",
  "main": "dist/types.cjs.js",
  "module": "dist/types.esm.js"
}
