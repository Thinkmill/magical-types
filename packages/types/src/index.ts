export type SignatureNode = {
  return: MagicalNode;
  parameters: Array<Parameter>;
  typeParameters: Array<TypeParameterNode>;
};

type IndexedAccessNode = {
  type: "IndexedAccess";
  object: MagicalNode;
  index: MagicalNode;
};

export type ObjectNode = {
  type: "Object";
  name: string | null;
  properties: Array<Property>;
  callSignatures: Array<SignatureNode>;
  constructSignatures: Array<SignatureNode>;
  aliasTypeArguments: Array<MagicalNode>;
};

export type ClassNode = {
  type: "Class";
  name: string | null;
  typeParameters: Array<MagicalNode>;
  thisNode: MagicalNode | null;
  properties: Array<Property>;
};

export type LazyNode = {
  type: "Lazy";
  loader: () => Promise<void> | void;
  value?: MagicalNode;
};

export type TypeParameterNode = { type: "TypeParameter"; value: string };

export type MagicalNode =
  | { type: "Error" }
  | { type: "Intrinsic"; value: string }
  | { type: "Promise"; value: MagicalNode }
  | { type: "StringLiteral"; value: string }
  | { type: "NumberLiteral"; value: number }
  | { type: "Array"; value: MagicalNode }
  | { type: "ReadonlyArray"; value: MagicalNode }
  | { type: "Union"; types: Array<MagicalNode>; name: string | null }
  | { type: "Intersection"; types: Array<MagicalNode> }
  | { type: "Symbol"; name: string }
  | {
      type: "Conditional";
      true: MagicalNode;
      false: MagicalNode;
      check: MagicalNode;
      extends: MagicalNode;
    }
  | TypeParameterNode
  | IndexedAccessNode
  | { type: "Tuple"; value: Array<MagicalNode> }
  | ClassNode
  | ObjectNode
  | LazyNode;

export type PositionedMagicalNode = {
  path: Array<string | number>;
  node: MagicalNode;
  depth: number;
};

export type Property = {
  key: string;
  description: string;
  value: MagicalNode;
  required: boolean;
};

export type Parameter = {
  name: string;
  type: MagicalNode;
  required: boolean;
};

export type MagicalNodeIndex = number & { __magicalNodeIndex: any };

type InnerReplace<Thing> = Thing extends MagicalNode
  ? MagicalNodeIndex
  : Thing extends object
  ? Thing extends LazyNode
    ? Thing
    : ReplaceMagicalNode<Thing>
  : Thing;

type ReplaceMagicalNode<Thing> = Thing extends LazyNode
  ? Thing
  : {
      [Key in keyof Thing]: Thing[Key] extends Array<infer Element>
        ? Array<InnerReplace<Element>>
        : InnerReplace<Thing[Key]>;
    };

export type MagicalNodeWithIndexes = ReplaceMagicalNode<MagicalNode>;
