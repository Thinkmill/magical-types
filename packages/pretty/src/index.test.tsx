import React from "react";
import { render } from "@testing-library/react";
import { Types } from "@magical-types/pretty";
import { ObjectNode } from "@magical-types/types";

let nodeWithCircularParam: ObjectNode = {
  type: "Object",
  aliasTypeArguments: [],
  callSignatures: [
    {
      parameters: [
        {
          name: "thing",
          required: true,
          type: { type: "Intrinsic", value: "string" }
        }
      ],
      return: { type: "Intrinsic", value: "string" },
      typeParameters: []
    }
  ],
  constructSignatures: [],
  name: "",
  properties: []
};

nodeWithCircularParam.callSignatures[0].parameters[0].type = nodeWithCircularParam;

let nodeWithCircularReturn: ObjectNode = {
  type: "Object",
  aliasTypeArguments: [],
  callSignatures: [
    {
      parameters: [
        {
          name: "thing",
          required: true,
          type: { type: "Intrinsic", value: "string" }
        }
      ],
      return: { type: "Intrinsic", value: "string" },
      typeParameters: []
    }
  ],
  constructSignatures: [],
  name: "",
  properties: []
};

nodeWithCircularReturn.callSignatures[0].return = nodeWithCircularReturn;

test("function with circular param", () => {
  let { container } = render(<Types node={nodeWithCircularParam} />);
  expect(container).toMatchSnapshot();
});

test("function with circular return", () => {
  let { container } = render(<Types node={nodeWithCircularReturn} />);
  expect(container).toMatchSnapshot();
});
