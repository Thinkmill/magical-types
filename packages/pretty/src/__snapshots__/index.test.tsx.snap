// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function with circular param 1`] = `
.emotion-5 {
  font-family: source-code-pro,Menlo,Monaco,Consolas,Courier New,monospace;
}

.emotion-0 {
  background-color: #F4F5F7;
  color: #5E6C84;
  border: 0;
  font-size: 14px;
  line-height: 20px;
  width: auto;
  margin: 2px 0;
  padding: 0 0.2em;
}

.emotion-0:hover,
.emotion-0:hover ~ .state-bit,
.state-bit ~ .emotion-0:hover {
  background-color: #6554C0;
  color: white;
}

.emotion-1 {
  color: #006644;
}

.emotion-3 {
  background-color: #EAE6FF;
  border-radius: 3px;
  color: #403294;
  display: inline-block;
  margin: 2px 0;
  padding: 0 0.2em;
}

<div>
  <div
    class="emotion-5"
  >
    <span>
      
      <button
        class="state-bit emotion-0"
        type="button"
      >
        (
        ...
        )
      </button>
      <span
        class="emotion-1"
      >
         =&gt; 
      </span>
      <button
        class="state-bit emotion-0"
        type="button"
      />
      <span
        class="emotion-3"
      >
        string
      </span>
      <button
        class="state-bit emotion-0"
        type="button"
      />
    </span>
  </div>
</div>
`;

exports[`function with circular return 1`] = `
.emotion-6 {
  font-family: source-code-pro,Menlo,Monaco,Consolas,Courier New,monospace;
}

.emotion-0 {
  background-color: #F4F5F7;
  color: #5E6C84;
  border: 0;
  font-size: 14px;
  line-height: 20px;
  width: auto;
  margin: 2px 0;
  padding: 0 0.2em;
}

.emotion-0:hover,
.emotion-0:hover ~ .state-bit,
.state-bit ~ .emotion-0:hover {
  background-color: #6554C0;
  color: white;
}

.emotion-4 {
  color: #006644;
}

.emotion-1 {
  background-color: #EAE6FF;
  border-radius: 3px;
  color: #403294;
  display: inline-block;
  margin: 2px 0;
  padding: 0 0.2em;
}

<div>
  <div
    class="emotion-6"
  >
    <span>
      
      <button
        class="state-bit emotion-0"
        type="button"
      >
        (
      </button>
      <span
        class="emotion-1"
      >
        thing
        
        :
         
      </span>
      <span
        class="emotion-1"
      >
        string
      </span>
      
      <button
        class="state-bit emotion-0"
        type="button"
      >
        )
      </button>
      <span
        class="emotion-4"
      >
         =&gt; 
      </span>
      <button
        class="state-bit emotion-0"
        type="button"
      >
        
        ...
        
      </button>
    </span>
  </div>
</div>
`;
